File: android/src/main/java/expo/modules/haishinkit/ExpoHaishinkitModule.kt
package expo.modules.haishinkit

import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition

class ExpoHaishinkitModule : Module() {
    override fun definition() = ModuleDefinition {
        Name("ExpoHaishinkit")

        // View definition with props and events
        View(ExpoHaishinkitView::class) {
            // Props
            Prop("url") { view: ExpoHaishinkitView, url: String ->
                view.url = url
            }

            Prop("streamName") { view: ExpoHaishinkitView, streamName: String ->
                view.streamName = streamName
            }

            Prop("camera") { view: ExpoHaishinkitView, camera: String ->
                if (camera == "front" || camera == "back") {
                    view.cameraPosition = camera
                    view.updateCamera()
                }
            }
            
            Prop("videoSettings") { view: ExpoHaishinkitView, settings: Map<String, Any?>? ->
                if (settings != null) {
                    view.setVideoSettings(settings)
                }
            }
            
            Prop("audioSettings") { view: ExpoHaishinkitView, settings: Map<String, Any?>? ->
                if (settings != null) {
                    view.setAudioSettings(settings)
                }
            }

            // Events
            Events(
                "onConnectionStatusChange",
                "onStreamStatusChange"
            )

            // Async functions for ref methods
            AsyncFunction("startPublishing") { view: ExpoHaishinkitView ->
                view.startPublishing()
            }

      AsyncFunction("stopPublishing") { view: ExpoHaishinkitView ->
        view.stopPublishing()
      }
        }
    }
}

File: ios/ExpoHaishinkitModule.swift
import ExpoModulesCore

public class ExpoHaishinkitModule: Module {
  public func definition() -> ModuleDefinition {
    Name("ExpoHaishinkit")

    // Enables the module to be used as a native view. Definition components that are accepted as part of the
    // view definition: Prop, Events.
    View(ExpoHaishinkitView.self) {
      // Defines a setter for the `url` prop.
      Prop("url") { (view: ExpoHaishinkitView, url: String?) in
        if let url = url {
          view.url = url
        }
      }
      
      Prop("streamName") { (view: ExpoHaishinkitView, streamName: String?) in
        if let streamName = streamName {
          view.streamName = streamName
        }
      }
      
      Prop("camera") { (view: ExpoHaishinkitView, camera: String?) in
        if let camera = camera, (camera == "front" || camera == "back") {
          view.camera = camera
          view.updateCamera()  // 카메라 변경
        }
      }
      
      Events(
        "onConnectionStatusChange",
        "onStreamStatusChange"
      )
      
      // AsyncFunction in View definition automatically receives view instance
      AsyncFunction("startPublishing") { (view: ExpoHaishinkitView) in
        view.startPublishing()
      }
      
      AsyncFunction("stopPublishing") { (view: ExpoHaishinkitView) in
        view.stopPublishing()
      }
    }
  }
}
